Source/WebCore/platform/graphics/BitmapImageMetadata.h

```diff
@@ -0,0 +1,117 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "Color.h"
+#include "DestinationColorSpace.h"
+#include "ImageOrientation.h"
+#include "ImageTypes.h"
+#include "IntPoint.h"
+#include <wtf/OptionSet.h>
+#include <wtf/ThreadSafeWeakPtr.h>
+
+namespace WebCore {
+
+class ImageSource;
+class ImageDecoder;
+class ImageFrame;
+class NativeImage;
+
+class BitmapImageMetadata {
+public:
+    BitmapImageMetadata(ImageSource&);
+
+    void clear() { m_cachedFlags = { }; }
+
+    EncodedDataStatus encodedDataStatus() const;
+    IntSize size() const;
+    std::optional<IntSize> densityCorrectedSize() const;
+    ImageOrientation orientation() const;
+    unsigned primaryFrameIndex() const;
+    unsigned frameCount() const;
+    RepetitionCount repetitionCount() const;
+    DestinationColorSpace colorSpace() const;
+    std::optional<Color> singlePixelSolidColor() const;
+
+    String uti() const;
+    String filenameExtension() const;
+    String accessibilityDescription() const;
+    std::optional<IntPoint> hotSpot() const;
+    SubsamplingLevel maximumSubsamplingLevel() const;
+
+    void dump(TextStream&) const;
+
+private:
+    enum class CachedFlag {
+        EncodedDataStatus           = 1 << 0,
+        Size                        = 1 << 1,
+        DensityCorrectedSize        = 1 << 2,
+        Orientation                 = 1 << 3,
+        PrimaryFrameIndex           = 1 << 4,
+        FrameCount                  = 1 << 5,
+        RepetitionCount             = 1 << 6,
+        ColorSpace                  = 1 << 7,
+        SinglePixelSolidColor       = 1 << 8,
+
+        UTI                         = 1 << 9,
+        FilenameExtension           = 1 << 10,
+        AccessibilityDescription    = 1 << 11,
+        HotSpot                     = 1 << 12,
+        MaximumSubsamplingLevel     = 1 << 13,
+    };
+
+    ImageSource& source() const { return *m_source.get(); }
```

Comment:
I think this should be named `protectedSource()` and return a RefPtr given that the `m_source` is a ThreadSafeWeakPtr and ThreadSafeWeakPtr::get() returns a smart pointer.

```
Ref<ImageSource> protectedSource() const { return m_source.get().releaseNonNull(); }
```